src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java

```diff
@@ -529,13 +528,27 @@ private String getProgrammingExerciseAllExerciseBuildsTriggeredTopic(Long progra
      * @param submission ProgrammingSubmission
      */
     public void notifyUserAboutSubmission(ProgrammingSubmission submission) {
-        String topic = Constants.PARTICIPATION_TOPIC_ROOT + submission.getParticipation().getId() + Constants.PROGRAMMING_SUBMISSION_TOPIC;
-        messagingTemplate.convertAndSend(topic, submission);
+        if (submission.getParticipation() instanceof StudentParticipation) {
+            StudentParticipation studentParticipation = (StudentParticipation) submission.getParticipation();
+            studentParticipation.getStudents().forEach(user -> messagingTemplate.convertAndSendToUser(user.getLogin(), NEW_SUBMISSION_TOPIC, submission));
+        }
+
+        if (submission.getParticipation() != null && submission.getParticipation().getExercise() != null) {
+            String exerciseTopic = EXERCISE_TOPIC_ROOT + submission.getParticipation().getExercise().getId() + "/newSubmission";
```

Comment:
Maybe there can be a static builder method that constructs this topic? At least `/newSubmission` should not be hardcoded and appear in multiple places.

Reply:
I've created a builder for it.